miner/Miner.hs

```diff
@@ -185,49 +176,40 @@ pCores = option auto
      <> help "Number of CPU cores to use (default: 1)")
 
 pVersion :: Parser ChainwebVersion
-pVersion = option cver
-    (long "version" <> metavar "VERSION"
-     <> value defv
+pVersion = textOption
+    (long "version" <> metavar "VERSION" <> value defv
      <> help ("Chainweb Network Version (default: " <> T.unpack (toText defv) <> ")"))
   where
     defv :: ChainwebVersion
     defv = Development
 
-    cver :: ReadM ChainwebVersion
-    cver = eitherReader $ \s ->
-        note "Illegal ChainwebVersion" . chainwebVersionFromText $ T.pack s
-
-pPort :: Parser Int
-pPort = option auto
-    (long "port" <> metavar "PORT" <> value 8081
-     <> help "Port on which to run the miner (default: 8081)")
-
-pUrl :: Parser URL
-pUrl = URL . hostAddressToBaseUrl Https <$> host
+pUrl :: Parser NodeURL
+pUrl = NodeURL . hostAddressToBaseUrl Https <$> hadd
   where
-    host :: Parser HostAddress
-    host = textOption
-        (long "node"
-        <> help "Remote address of Chainweb Node to send mining results to."
-        <> metavar "<HOSTNAME:PORT>")
+    hadd :: Parser HostAddress
+    hadd = textOption
+        (long "node" <> metavar "<HOSTNAME:PORT>"
+        <> help "Remote address of Chainweb Node to send mining results to.")
+
+pChainId :: Parser (Maybe ChainId)
+pChainId = optional $ textOption
+    (long "chain" <> metavar "CHAIN-ID"
+     <> help "Prioritize work requests for a specific chain")
 
 --------------------------------------------------------------------------------
 -- Work
 
 main :: IO ()
 main = do
-    env@(Env _ _ _ as) <- Env
-        <$> newEmptyTMVarIO
-        <*> MWC.createSystemRandom
+    env@(Env _ _ as) <- Env
+        <$> MWC.createSystemRandom
         <*> newManager (mkManagerSettings ss Nothing)
         <*> execParser opts
-    case cmd as of
-        GPU _ -> putStrLn "GPU mining is not yet available."
-        CPU _ -> do
-            miner <- async $ mining (scheme env) env
-            pPrintNoColor as
-            W.run (port as) $ app env
-            wait miner
+    getWork env >>= \case
+        Nothing -> putStrLn "Failed to connect to the given Chainweb Node." >> exitFailure
+        Just wb -> case cmd as of
+            GPU _ -> putStrLn "GPU mining is not yet available."
+            CPU _ -> pPrintNoColor as >> mining (scheme env) env wb >> exitFailure
```

Comment:
The re-echo of the arguments may not be useful.

Prod-level logging hasn't yet been addressed at all, but will be done so in a follow-up.